src/core/geometry/qgslinestring.cpp

```diff
@@ -2817,3 +2817,44 @@ std::unique_ptr< QgsLineString > QgsLineString::interpolateM( bool use3DDistance
   }
   return std::make_unique< QgsLineString >( xOut, yOut, zOut, mOut );
 }
+
+double QgsLineString::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for linestrings (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  if ( fromVertexNumber < 0 || fromVertexNumber >= numPoints() || toVertexNumber < 0 || toVertexNumber >= numPoints() )
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Ensure fromVertex < toVertex for simplicity
+  if ( fromVertexNumber > toVertexNumber )
+  {
+    return distanceBetweenVertices( QgsVertexId( 0, 0, toVertexNumber ), QgsVertexId( 0, 0, fromVertexNumber ) );
+  }
+
+  double totalDistance = 0.0;
+
+  // For linestring, just accumulate Euclidean distances between consecutive points
+  for ( int i = fromVertexNumber; i < toVertexNumber; ++i )
+  {
+    double dx = mX[i + 1] - mX[i];
+    double dy = mY[i + 1] - mY[i];
+    double dz = 0.0;
+
+    if ( is3D() && i + 1 < mZ.size() )
```

Comment:
retrieve this once and store in a bool outside of the loop -- calling is3D() is not free.

Reply:
done